NPE while Mapping Fragments

@sav007 We've recently been getting a `NullPointerException` in a generated query when mapping to a Fragment.  The generated code is calling
```Utils.checkNotNull(ourFragment, "ourFragment == null")```
but the Fragments are marked as nullable.

```
Caused by: com.apollographql.apollo.exception.ApolloParseException: Failed to parse http response
                         E      at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:102)
                         E      at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.intercept(ApolloParseInterceptor.java:51)
                         E      at com.apollographql.apollo.internal.interceptor.RealApolloInterceptorChain.proceed(RealApolloInterceptorChain.java:38)
                         E      at com.apollographql.apollo.internal.interceptor.ApolloCacheInterceptor.intercept(ApolloCacheInterceptor.java:65)
                         E      at com.apollographql.apollo.internal.interceptor.RealApolloInterceptorChain.proceed(RealApolloInterceptorChain.java:38)
                         E      at com.apollographql.apollo.internal.fetcher.CacheFirstFetcher$CacheFirstInterceptor.intercept(CacheFirstFetcher.java:40)
                         E      at com.apollographql.apollo.internal.interceptor.RealApolloInterceptorChain.proceed(RealApolloInterceptorChain.java:38)
                         E      at com.apollographql.apollo.internal.RealApolloCall.execute(RealApolloCall.java:130)
```

I'll DM you with more details

--- a/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/FragmentsResponseMapperBuilder.kt
+++ b/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/FragmentsResponseMapperBuilder.kt
@@ -85,7 +85,7 @@ class FragmentsResponseMapperBuilder(
           .add("return new \$L(", SchemaTypeSpecBuilder.FRAGMENTS_FIELD.type.withoutAnnotations())
           .add(fragmentFields
               .mapIndexed { i, fieldSpec ->
-                if (fieldSpec.type.isOptional()) {
+                if (fieldSpec.type.isNullable()) {
                   CodeBlock.of("\$L\$L", if (i > 0) ", " else "", fieldSpec.name)
                 } else {
                   CodeBlock.of("\$L\$T.checkNotNull(\$L, \$S)", if (i > 0) ", " else "", ClassNames.API_UTILS,

--- a/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Util.kt
+++ b/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Util.kt
@@ -2,7 +2,6 @@ package com.apollographql.apollo.compiler
 
 import com.apollographql.apollo.compiler.ir.CodeGenerationContext
 import com.squareup.javapoet.*
-import java.lang.NullPointerException
 import javax.lang.model.element.Modifier
 
 private val JAVA_RESERVED_WORDS = arrayOf(
@@ -236,6 +235,8 @@ fun TypeSpec.withHashCodeImplementation(): TypeSpec {
 }
 
 fun ClassName.mapperFieldName(): String = "${simpleName().decapitalize()}FieldMapper"
+
+fun TypeName.isNullable(): Boolean = isOptional() || annotations.contains(Annotations.NULLABLE)
 
 fun TypeName.isOptional(): Boolean {
   val rawType = (this as? ParameterizedTypeName)?.rawType ?: this

--- a/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/TestOperation.graphql
+++ b/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/TestOperation.graphql
@@ -1,0 +1,20 @@
+query TestQuery {
+  r2: hero {
+    ...HumanDetails
+    ...DroidDetails
+  }
+  luke: hero {
+    ...HumanDetails
+    ...DroidDetails
+  }
+}
+
+fragment HumanDetails on Human {
+  name
+  height
+}
+
+fragment DroidDetails on Droid {
+  name
+  primaryFunction
+}

--- a/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/TestOperation.json
+++ b/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/TestOperation.json
@@ -1,0 +1,131 @@
+{
+	"operations": [
+		{
+			"filePath": "src/test/graphql/com/example/fragments_with_type_condition_nullable/TestOperation.graphql",
+			"operationName": "TestQuery",
+			"operationType": "query",
+			"rootType": "Query",
+			"variables": [],
+			"source": "query TestQuery {\n  r2: hero {\n    __typename\n    ...HumanDetails\n    ...DroidDetails\n  }\n  luke: hero {\n    __typename\n    ...HumanDetails\n    ...DroidDetails\n  }\n}",
+			"fields": [
+				{
+					"responseName": "r2",
+					"fieldName": "hero",
+					"type": "Character",
+					"isDeprecated": false,
+					"deprecationReason": null,
+					"fields": [
+						{
+							"responseName": "__typename",
+							"fieldName": "__typename",
+							"type": "String!"
+						}
+					],
+					"fragmentSpreads": [
+						"HumanDetails",
+						"DroidDetails"
+					],
+					"inlineFragments": []
+				},
+				{
+					"responseName": "luke",
+					"fieldName": "hero",
+					"type": "Character",
+					"isDeprecated": false,
+					"deprecationReason": null,
+					"fields": [
+						{
+							"responseName": "__typename",
+							"fieldName": "__typename",
+							"type": "String!"
+						}
+					],
+					"fragmentSpreads": [
+						"HumanDetails",
+						"DroidDetails"
+					],
+					"inlineFragments": []
+				}
+			],
+			"fragmentsReferenced": [
+				"HumanDetails",
+				"DroidDetails"
+			],
+			"sourceWithFragments": "query TestQuery {\n  r2: hero {\n    __typename\n    ...HumanDetails\n    ...DroidDetails\n  }\n  luke: hero {\n    __typename\n    ...HumanDetails\n    ...DroidDetails\n  }\n}\nfragment DroidDetails on Droid {\n  __typename\n  name\n  primaryFunction\n}\nfragment HumanDetails on Human {\n  __typename\n  name\n  height\n}",
+			"operationId": "185ee12f775bf02624bb5f646f5ed2de3009860b79380264ce4716e65fba947d"
+		}
+	],
+	"fragments": [
+		{
+			"filePath": "src/test/graphql/com/example/fragments_with_type_condition/TestOperation.graphql",
+			"fragmentName": "HumanDetails",
+			"source": "fragment HumanDetails on Human {\n  __typename\n  name\n  height\n}",
+			"typeCondition": "Human",
+			"possibleTypes": [
+				"Human"
+			],
+			"fields": [
+				{
+					"responseName": "__typename",
+					"fieldName": "__typename",
+					"type": "String!"
+				},
+				{
+					"responseName": "name",
+					"fieldName": "name",
+					"type": "String!",
+					"description": "What this human calls themselves",
+					"isDeprecated": false,
+					"deprecationReason": null
+				},
+				{
+					"responseName": "height",
+					"fieldName": "height",
+					"type": "Float",
+					"description": "Height in the preferred unit, default is meters",
+					"isDeprecated": false,
+					"deprecationReason": null
+				}
+			],
+			"fragmentSpreads": [],
+			"inlineFragments": [],
+			"fragmentsReferenced": []
+		},
+		{
+			"filePath": "src/test/graphql/com/example/fragments_with_type_condition/TestOperation.graphql",
+			"fragmentName": "DroidDetails",
+			"source": "fragment DroidDetails on Droid {\n  __typename\n  name\n  primaryFunction\n}",
+			"typeCondition": "Droid",
+			"possibleTypes": [
+				"Droid"
+			],
+			"fields": [
+				{
+					"responseName": "__typename",
+					"fieldName": "__typename",
+					"type": "String!"
+				},
+				{
+					"responseName": "name",
+					"fieldName": "name",
+					"type": "String!",
+					"description": "What others call this droid",
+					"isDeprecated": false,
+					"deprecationReason": null
+				},
+				{
+					"responseName": "primaryFunction",
+					"fieldName": "primaryFunction",
+					"type": "String",
+					"description": "This droid's primary function",
+					"isDeprecated": false,
+					"deprecationReason": null
+				}
+			],
+			"fragmentSpreads": [],
+			"inlineFragments": [],
+			"fragmentsReferenced": []
+		}
+	],
+	"typesUsed": []
+}

--- a/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/TestQuery.java
+++ b/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/TestQuery.java
@@ -1,0 +1,584 @@
+package com.example.fragments_with_type_condition_nullable;
+
+import com.apollographql.apollo.api.FragmentResponseFieldMapper;
+import com.apollographql.apollo.api.Operation;
+import com.apollographql.apollo.api.OperationName;
+import com.apollographql.apollo.api.Query;
+import com.apollographql.apollo.api.ResponseField;
+import com.apollographql.apollo.api.ResponseFieldMapper;
+import com.apollographql.apollo.api.ResponseFieldMarshaller;
+import com.apollographql.apollo.api.ResponseReader;
+import com.apollographql.apollo.api.ResponseWriter;
+import com.apollographql.apollo.api.internal.Utils;
+import com.example.fragments_with_type_condition_nullable.fragment.DroidDetails;
+import com.example.fragments_with_type_condition_nullable.fragment.HumanDetails;
+import java.lang.Object;
+import java.lang.Override;
+import java.lang.String;
+import java.util.Arrays;
+import javax.annotation.Generated;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+@Generated("Apollo GraphQL")
+public final class TestQuery implements Query<TestQuery.Data, TestQuery.Data, Operation.Variables> {
+  public static final String OPERATION_DEFINITION = "query TestQuery {\n"
+      + "  r2: hero {\n"
+      + "    __typename\n"
+      + "    ...HumanDetails\n"
+      + "    ...DroidDetails\n"
+      + "  }\n"
+      + "  luke: hero {\n"
+      + "    __typename\n"
+      + "    ...HumanDetails\n"
+      + "    ...DroidDetails\n"
+      + "  }\n"
+      + "}";
+
+  public static final String QUERY_DOCUMENT = OPERATION_DEFINITION + "\n"
+   + HumanDetails.FRAGMENT_DEFINITION + "\n"
+   + DroidDetails.FRAGMENT_DEFINITION;
+
+  private static final OperationName OPERATION_NAME = new OperationName() {
+    @Override
+    public String name() {
+      return "TestQuery";
+    }
+  };
+
+  private final Operation.Variables variables;
+
+  public TestQuery() {
+    this.variables = Operation.EMPTY_VARIABLES;
+  }
+
+  @Override
+  public String operationId() {
+    return "185ee12f775bf02624bb5f646f5ed2de3009860b79380264ce4716e65fba947d";
+  }
+
+  @Override
+  public String queryDocument() {
+    return QUERY_DOCUMENT;
+  }
+
+  @Override
+  public TestQuery.Data wrapData(TestQuery.Data data) {
+    return data;
+  }
+
+  @Override
+  public Operation.Variables variables() {
+    return variables;
+  }
+
+  @Override
+  public ResponseFieldMapper<TestQuery.Data> responseFieldMapper() {
+    return new Data.Mapper();
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  @Override
+  public OperationName name() {
+    return OPERATION_NAME;
+  }
+
+  public static final class Builder {
+    Builder() {
+    }
+
+    public TestQuery build() {
+      return new TestQuery();
+    }
+  }
+
+  public static class Data implements Operation.Data {
+    static final ResponseField[] $responseFields = {
+      ResponseField.forObject("r2", "hero", null, true),
+      ResponseField.forObject("luke", "hero", null, true)
+    };
+
+    final @Nullable R2 r2;
+
+    final @Nullable Luke luke;
+
+    private volatile String $toString;
+
+    private volatile int $hashCode;
+
+    private volatile boolean $hashCodeMemoized;
+
+    public Data(@Nullable R2 r2, @Nullable Luke luke) {
+      this.r2 = r2;
+      this.luke = luke;
+    }
+
+    public @Nullable R2 r2() {
+      return this.r2;
+    }
+
+    public @Nullable Luke luke() {
+      return this.luke;
+    }
+
+    public ResponseFieldMarshaller marshaller() {
+      return new ResponseFieldMarshaller() {
+        @Override
+        public void marshal(ResponseWriter writer) {
+          writer.writeObject($responseFields[0], r2 != null ? r2.marshaller() : null);
+          writer.writeObject($responseFields[1], luke != null ? luke.marshaller() : null);
+        }
+      };
+    }
+
+    @Override
+    public String toString() {
+      if ($toString == null) {
+        $toString = "Data{"
+          + "r2=" + r2 + ", "
+          + "luke=" + luke
+          + "}";
+      }
+      return $toString;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (o instanceof Data) {
+        Data that = (Data) o;
+        return ((this.r2 == null) ? (that.r2 == null) : this.r2.equals(that.r2))
+         && ((this.luke == null) ? (that.luke == null) : this.luke.equals(that.luke));
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      if (!$hashCodeMemoized) {
+        int h = 1;
+        h *= 1000003;
+        h ^= (r2 == null) ? 0 : r2.hashCode();
+        h *= 1000003;
+        h ^= (luke == null) ? 0 : luke.hashCode();
+        $hashCode = h;
+        $hashCodeMemoized = true;
+      }
+      return $hashCode;
+    }
+
+    public static final class Mapper implements ResponseFieldMapper<Data> {
+      final R2.Mapper r2FieldMapper = new R2.Mapper();
+
+      final Luke.Mapper lukeFieldMapper = new Luke.Mapper();
+
+      @Override
+      public Data map(ResponseReader reader) {
+        final R2 r2 = reader.readObject($responseFields[0], new ResponseReader.ObjectReader<R2>() {
+          @Override
+          public R2 read(ResponseReader reader) {
+            return r2FieldMapper.map(reader);
+          }
+        });
+        final Luke luke = reader.readObject($responseFields[1], new ResponseReader.ObjectReader<Luke>() {
+          @Override
+          public Luke read(ResponseReader reader) {
+            return lukeFieldMapper.map(reader);
+          }
+        });
+        return new Data(r2, luke);
+      }
+    }
+  }
+
+  public static class R2 {
+    static final ResponseField[] $responseFields = {
+      ResponseField.forString("__typename", "__typename", null, false),
+      ResponseField.forFragment("__typename", "__typename", Arrays.asList("Human",
+      "Droid"))
+    };
+
+    final @Nonnull String __typename;
+
+    private final @Nonnull Fragments fragments;
+
+    private volatile String $toString;
+
+    private volatile int $hashCode;
+
+    private volatile boolean $hashCodeMemoized;
+
+    public R2(@Nonnull String __typename, @Nonnull Fragments fragments) {
+      this.__typename = Utils.checkNotNull(__typename, "__typename == null");
+      this.fragments = Utils.checkNotNull(fragments, "fragments == null");
+    }
+
+    public @Nonnull String __typename() {
+      return this.__typename;
+    }
+
+    public @Nonnull Fragments fragments() {
+      return this.fragments;
+    }
+
+    public ResponseFieldMarshaller marshaller() {
+      return new ResponseFieldMarshaller() {
+        @Override
+        public void marshal(ResponseWriter writer) {
+          writer.writeString($responseFields[0], __typename);
+          fragments.marshaller().marshal(writer);
+        }
+      };
+    }
+
+    @Override
+    public String toString() {
+      if ($toString == null) {
+        $toString = "R2{"
+          + "__typename=" + __typename + ", "
+          + "fragments=" + fragments
+          + "}";
+      }
+      return $toString;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (o instanceof R2) {
+        R2 that = (R2) o;
+        return this.__typename.equals(that.__typename)
+         && this.fragments.equals(that.fragments);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      if (!$hashCodeMemoized) {
+        int h = 1;
+        h *= 1000003;
+        h ^= __typename.hashCode();
+        h *= 1000003;
+        h ^= fragments.hashCode();
+        $hashCode = h;
+        $hashCodeMemoized = true;
+      }
+      return $hashCode;
+    }
+
+    public static class Fragments {
+      final @Nullable HumanDetails humanDetails;
+
+      final @Nullable DroidDetails droidDetails;
+
+      private volatile String $toString;
+
+      private volatile int $hashCode;
+
+      private volatile boolean $hashCodeMemoized;
+
+      public Fragments(@Nullable HumanDetails humanDetails, @Nullable DroidDetails droidDetails) {
+        this.humanDetails = humanDetails;
+        this.droidDetails = droidDetails;
+      }
+
+      public @Nullable HumanDetails humanDetails() {
+        return this.humanDetails;
+      }
+
+      public @Nullable DroidDetails droidDetails() {
+        return this.droidDetails;
+      }
+
+      public ResponseFieldMarshaller marshaller() {
+        return new ResponseFieldMarshaller() {
+          @Override
+          public void marshal(ResponseWriter writer) {
+            final HumanDetails $humanDetails = humanDetails;
+            if ($humanDetails != null) {
+              $humanDetails.marshaller().marshal(writer);
+            }
+            final DroidDetails $droidDetails = droidDetails;
+            if ($droidDetails != null) {
+              $droidDetails.marshaller().marshal(writer);
+            }
+          }
+        };
+      }
+
+      @Override
+      public String toString() {
+        if ($toString == null) {
+          $toString = "Fragments{"
+            + "humanDetails=" + humanDetails + ", "
+            + "droidDetails=" + droidDetails
+            + "}";
+        }
+        return $toString;
+      }
+
+      @Override
+      public boolean equals(Object o) {
+        if (o == this) {
+          return true;
+        }
+        if (o instanceof Fragments) {
+          Fragments that = (Fragments) o;
+          return ((this.humanDetails == null) ? (that.humanDetails == null) : this.humanDetails.equals(that.humanDetails))
+           && ((this.droidDetails == null) ? (that.droidDetails == null) : this.droidDetails.equals(that.droidDetails));
+        }
+        return false;
+      }
+
+      @Override
+      public int hashCode() {
+        if (!$hashCodeMemoized) {
+          int h = 1;
+          h *= 1000003;
+          h ^= (humanDetails == null) ? 0 : humanDetails.hashCode();
+          h *= 1000003;
+          h ^= (droidDetails == null) ? 0 : droidDetails.hashCode();
+          $hashCode = h;
+          $hashCodeMemoized = true;
+        }
+        return $hashCode;
+      }
+
+      public static final class Mapper implements FragmentResponseFieldMapper<Fragments> {
+        final HumanDetails.Mapper humanDetailsFieldMapper = new HumanDetails.Mapper();
+
+        final DroidDetails.Mapper droidDetailsFieldMapper = new DroidDetails.Mapper();
+
+        @Override
+        public @Nonnull Fragments map(ResponseReader reader, @Nonnull String conditionalType) {
+          HumanDetails humanDetails = null;
+          DroidDetails droidDetails = null;
+          if (HumanDetails.POSSIBLE_TYPES.contains(conditionalType)) {
+            humanDetails = humanDetailsFieldMapper.map(reader);
+          }
+          if (DroidDetails.POSSIBLE_TYPES.contains(conditionalType)) {
+            droidDetails = droidDetailsFieldMapper.map(reader);
+          }
+          return new Fragments(humanDetails, droidDetails);
+        }
+      }
+    }
+
+    public static final class Mapper implements ResponseFieldMapper<R2> {
+      final Fragments.Mapper fragmentsFieldMapper = new Fragments.Mapper();
+
+      @Override
+      public R2 map(ResponseReader reader) {
+        final String __typename = reader.readString($responseFields[0]);
+        final Fragments fragments = reader.readConditional((ResponseField.ConditionalTypeField) $responseFields[1], new ResponseReader.ConditionalTypeReader<Fragments>() {
+          @Override
+          public Fragments read(String conditionalType, ResponseReader reader) {
+            return fragmentsFieldMapper.map(reader, conditionalType);
+          }
+        });
+        return new R2(__typename, fragments);
+      }
+    }
+  }
+
+  public static class Luke {
+    static final ResponseField[] $responseFields = {
+      ResponseField.forString("__typename", "__typename", null, false),
+      ResponseField.forFragment("__typename", "__typename", Arrays.asList("Human",
+      "Droid"))
+    };
+
+    final @Nonnull String __typename;
+
+    private final @Nonnull Fragments fragments;
+
+    private volatile String $toString;
+
+    private volatile int $hashCode;
+
+    private volatile boolean $hashCodeMemoized;
+
+    public Luke(@Nonnull String __typename, @Nonnull Fragments fragments) {
+      this.__typename = Utils.checkNotNull(__typename, "__typename == null");
+      this.fragments = Utils.checkNotNull(fragments, "fragments == null");
+    }
+
+    public @Nonnull String __typename() {
+      return this.__typename;
+    }
+
+    public @Nonnull Fragments fragments() {
+      return this.fragments;
+    }
+
+    public ResponseFieldMarshaller marshaller() {
+      return new ResponseFieldMarshaller() {
+        @Override
+        public void marshal(ResponseWriter writer) {
+          writer.writeString($responseFields[0], __typename);
+          fragments.marshaller().marshal(writer);
+        }
+      };
+    }
+
+    @Override
+    public String toString() {
+      if ($toString == null) {
+        $toString = "Luke{"
+          + "__typename=" + __typename + ", "
+          + "fragments=" + fragments
+          + "}";
+      }
+      return $toString;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (o instanceof Luke) {
+        Luke that = (Luke) o;
+        return this.__typename.equals(that.__typename)
+         && this.fragments.equals(that.fragments);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      if (!$hashCodeMemoized) {
+        int h = 1;
+        h *= 1000003;
+        h ^= __typename.hashCode();
+        h *= 1000003;
+        h ^= fragments.hashCode();
+        $hashCode = h;
+        $hashCodeMemoized = true;
+      }
+      return $hashCode;
+    }
+
+    public static class Fragments {
+      final @Nullable HumanDetails humanDetails;
+
+      final @Nullable DroidDetails droidDetails;
+
+      private volatile String $toString;
+
+      private volatile int $hashCode;
+
+      private volatile boolean $hashCodeMemoized;
+
+      public Fragments(@Nullable HumanDetails humanDetails, @Nullable DroidDetails droidDetails) {
+        this.humanDetails = humanDetails;
+        this.droidDetails = droidDetails;
+      }
+
+      public @Nullable HumanDetails humanDetails() {
+        return this.humanDetails;
+      }
+
+      public @Nullable DroidDetails droidDetails() {
+        return this.droidDetails;
+      }
+
+      public ResponseFieldMarshaller marshaller() {
+        return new ResponseFieldMarshaller() {
+          @Override
+          public void marshal(ResponseWriter writer) {
+            final HumanDetails $humanDetails = humanDetails;
+            if ($humanDetails != null) {
+              $humanDetails.marshaller().marshal(writer);
+            }
+            final DroidDetails $droidDetails = droidDetails;
+            if ($droidDetails != null) {
+              $droidDetails.marshaller().marshal(writer);
+            }
+          }
+        };
+      }
+
+      @Override
+      public String toString() {
+        if ($toString == null) {
+          $toString = "Fragments{"
+            + "humanDetails=" + humanDetails + ", "
+            + "droidDetails=" + droidDetails
+            + "}";
+        }
+        return $toString;
+      }
+
+      @Override
+      public boolean equals(Object o) {
+        if (o == this) {
+          return true;
+        }
+        if (o instanceof Fragments) {
+          Fragments that = (Fragments) o;
+          return ((this.humanDetails == null) ? (that.humanDetails == null) : this.humanDetails.equals(that.humanDetails))
+           && ((this.droidDetails == null) ? (that.droidDetails == null) : this.droidDetails.equals(that.droidDetails));
+        }
+        return false;
+      }
+
+      @Override
+      public int hashCode() {
+        if (!$hashCodeMemoized) {
+          int h = 1;
+          h *= 1000003;
+          h ^= (humanDetails == null) ? 0 : humanDetails.hashCode();
+          h *= 1000003;
+          h ^= (droidDetails == null) ? 0 : droidDetails.hashCode();
+          $hashCode = h;
+          $hashCodeMemoized = true;
+        }
+        return $hashCode;
+      }
+
+      public static final class Mapper implements FragmentResponseFieldMapper<Fragments> {
+        final HumanDetails.Mapper humanDetailsFieldMapper = new HumanDetails.Mapper();
+
+        final DroidDetails.Mapper droidDetailsFieldMapper = new DroidDetails.Mapper();
+
+        @Override
+        public @Nonnull Fragments map(ResponseReader reader, @Nonnull String conditionalType) {
+          HumanDetails humanDetails = null;
+          DroidDetails droidDetails = null;
+          if (HumanDetails.POSSIBLE_TYPES.contains(conditionalType)) {
+            humanDetails = humanDetailsFieldMapper.map(reader);
+          }
+          if (DroidDetails.POSSIBLE_TYPES.contains(conditionalType)) {
+            droidDetails = droidDetailsFieldMapper.map(reader);
+          }
+          return new Fragments(humanDetails, droidDetails);
+        }
+      }
+    }
+
+    public static final class Mapper implements ResponseFieldMapper<Luke> {
+      final Fragments.Mapper fragmentsFieldMapper = new Fragments.Mapper();
+
+      @Override
+      public Luke map(ResponseReader reader) {
+        final String __typename = reader.readString($responseFields[0]);
+        final Fragments fragments = reader.readConditional((ResponseField.ConditionalTypeField) $responseFields[1], new ResponseReader.ConditionalTypeReader<Fragments>() {
+          @Override
+          public Fragments read(String conditionalType, ResponseReader reader) {
+            return fragmentsFieldMapper.map(reader, conditionalType);
+          }
+        });
+        return new Luke(__typename, fragments);
+      }
+    }
+  }
+}

--- a/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/fragment/DroidDetails.java
+++ b/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/fragment/DroidDetails.java
@@ -1,0 +1,135 @@
+package com.example.fragments_with_type_condition_nullable.fragment;
+
+import com.apollographql.apollo.api.GraphqlFragment;
+import com.apollographql.apollo.api.ResponseField;
+import com.apollographql.apollo.api.ResponseFieldMapper;
+import com.apollographql.apollo.api.ResponseFieldMarshaller;
+import com.apollographql.apollo.api.ResponseReader;
+import com.apollographql.apollo.api.ResponseWriter;
+import com.apollographql.apollo.api.internal.Utils;
+import java.lang.Object;
+import java.lang.Override;
+import java.lang.String;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Generated;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+@Generated("Apollo GraphQL")
+public class DroidDetails implements GraphqlFragment {
+  static final ResponseField[] $responseFields = {
+    ResponseField.forString("__typename", "__typename", null, false),
+    ResponseField.forString("name", "name", null, false),
+    ResponseField.forString("primaryFunction", "primaryFunction", null, true)
+  };
+
+  public static final String FRAGMENT_DEFINITION = "fragment DroidDetails on Droid {\n"
+      + "  __typename\n"
+      + "  name\n"
+      + "  primaryFunction\n"
+      + "}";
+
+  public static final List<String> POSSIBLE_TYPES = Collections.unmodifiableList(Arrays.asList( "Droid"));
+
+  final @Nonnull String __typename;
+
+  final @Nonnull String name;
+
+  final @Nullable String primaryFunction;
+
+  private volatile String $toString;
+
+  private volatile int $hashCode;
+
+  private volatile boolean $hashCodeMemoized;
+
+  public DroidDetails(@Nonnull String __typename, @Nonnull String name,
+      @Nullable String primaryFunction) {
+    this.__typename = Utils.checkNotNull(__typename, "__typename == null");
+    this.name = Utils.checkNotNull(name, "name == null");
+    this.primaryFunction = primaryFunction;
+  }
+
+  public @Nonnull String __typename() {
+    return this.__typename;
+  }
+
+  /**
+   * What others call this droid
+   */
+  public @Nonnull String name() {
+    return this.name;
+  }
+
+  /**
+   * This droid's primary function
+   */
+  public @Nullable String primaryFunction() {
+    return this.primaryFunction;
+  }
+
+  public ResponseFieldMarshaller marshaller() {
+    return new ResponseFieldMarshaller() {
+      @Override
+      public void marshal(ResponseWriter writer) {
+        writer.writeString($responseFields[0], __typename);
+        writer.writeString($responseFields[1], name);
+        writer.writeString($responseFields[2], primaryFunction);
+      }
+    };
+  }
+
+  @Override
+  public String toString() {
+    if ($toString == null) {
+      $toString = "DroidDetails{"
+        + "__typename=" + __typename + ", "
+        + "name=" + name + ", "
+        + "primaryFunction=" + primaryFunction
+        + "}";
+    }
+    return $toString;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (o == this) {
+      return true;
+    }
+    if (o instanceof DroidDetails) {
+      DroidDetails that = (DroidDetails) o;
+      return this.__typename.equals(that.__typename)
+       && this.name.equals(that.name)
+       && ((this.primaryFunction == null) ? (that.primaryFunction == null) : this.primaryFunction.equals(that.primaryFunction));
+    }
+    return false;
+  }
+
+  @Override
+  public int hashCode() {
+    if (!$hashCodeMemoized) {
+      int h = 1;
+      h *= 1000003;
+      h ^= __typename.hashCode();
+      h *= 1000003;
+      h ^= name.hashCode();
+      h *= 1000003;
+      h ^= (primaryFunction == null) ? 0 : primaryFunction.hashCode();
+      $hashCode = h;
+      $hashCodeMemoized = true;
+    }
+    return $hashCode;
+  }
+
+  public static final class Mapper implements ResponseFieldMapper<DroidDetails> {
+    @Override
+    public DroidDetails map(ResponseReader reader) {
+      final String __typename = reader.readString($responseFields[0]);
+      final String name = reader.readString($responseFields[1]);
+      final String primaryFunction = reader.readString($responseFields[2]);
+      return new DroidDetails(__typename, name, primaryFunction);
+    }
+  }
+}

--- a/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/fragment/HumanDetails.java
+++ b/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/fragment/HumanDetails.java
@@ -1,0 +1,135 @@
+package com.example.fragments_with_type_condition_nullable.fragment;
+
+import com.apollographql.apollo.api.GraphqlFragment;
+import com.apollographql.apollo.api.ResponseField;
+import com.apollographql.apollo.api.ResponseFieldMapper;
+import com.apollographql.apollo.api.ResponseFieldMarshaller;
+import com.apollographql.apollo.api.ResponseReader;
+import com.apollographql.apollo.api.ResponseWriter;
+import com.apollographql.apollo.api.internal.Utils;
+import java.lang.Double;
+import java.lang.Object;
+import java.lang.Override;
+import java.lang.String;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Generated;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+@Generated("Apollo GraphQL")
+public class HumanDetails implements GraphqlFragment {
+  static final ResponseField[] $responseFields = {
+    ResponseField.forString("__typename", "__typename", null, false),
+    ResponseField.forString("name", "name", null, false),
+    ResponseField.forDouble("height", "height", null, true)
+  };
+
+  public static final String FRAGMENT_DEFINITION = "fragment HumanDetails on Human {\n"
+      + "  __typename\n"
+      + "  name\n"
+      + "  height\n"
+      + "}";
+
+  public static final List<String> POSSIBLE_TYPES = Collections.unmodifiableList(Arrays.asList( "Human"));
+
+  final @Nonnull String __typename;
+
+  final @Nonnull String name;
+
+  final @Nullable Double height;
+
+  private volatile String $toString;
+
+  private volatile int $hashCode;
+
+  private volatile boolean $hashCodeMemoized;
+
+  public HumanDetails(@Nonnull String __typename, @Nonnull String name, @Nullable Double height) {
+    this.__typename = Utils.checkNotNull(__typename, "__typename == null");
+    this.name = Utils.checkNotNull(name, "name == null");
+    this.height = height;
+  }
+
+  public @Nonnull String __typename() {
+    return this.__typename;
+  }
+
+  /**
+   * What this human calls themselves
+   */
+  public @Nonnull String name() {
+    return this.name;
+  }
+
+  /**
+   * Height in the preferred unit, default is meters
+   */
+  public @Nullable Double height() {
+    return this.height;
+  }
+
+  public ResponseFieldMarshaller marshaller() {
+    return new ResponseFieldMarshaller() {
+      @Override
+      public void marshal(ResponseWriter writer) {
+        writer.writeString($responseFields[0], __typename);
+        writer.writeString($responseFields[1], name);
+        writer.writeDouble($responseFields[2], height);
+      }
+    };
+  }
+
+  @Override
+  public String toString() {
+    if ($toString == null) {
+      $toString = "HumanDetails{"
+        + "__typename=" + __typename + ", "
+        + "name=" + name + ", "
+        + "height=" + height
+        + "}";
+    }
+    return $toString;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (o == this) {
+      return true;
+    }
+    if (o instanceof HumanDetails) {
+      HumanDetails that = (HumanDetails) o;
+      return this.__typename.equals(that.__typename)
+       && this.name.equals(that.name)
+       && ((this.height == null) ? (that.height == null) : this.height.equals(that.height));
+    }
+    return false;
+  }
+
+  @Override
+  public int hashCode() {
+    if (!$hashCodeMemoized) {
+      int h = 1;
+      h *= 1000003;
+      h ^= __typename.hashCode();
+      h *= 1000003;
+      h ^= name.hashCode();
+      h *= 1000003;
+      h ^= (height == null) ? 0 : height.hashCode();
+      $hashCode = h;
+      $hashCodeMemoized = true;
+    }
+    return $hashCode;
+  }
+
+  public static final class Mapper implements ResponseFieldMapper<HumanDetails> {
+    @Override
+    public HumanDetails map(ResponseReader reader) {
+      final String __typename = reader.readString($responseFields[0]);
+      final String name = reader.readString($responseFields[1]);
+      final Double height = reader.readDouble($responseFields[2]);
+      return new HumanDetails(__typename, name, height);
+    }
+  }
+}

--- a/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/type/CustomType.java
+++ b/apollo-compiler/src/test/graphql/com/example/fragments_with_type_condition_nullable/type/CustomType.java
@@ -1,0 +1,22 @@
+package com.example.fragments_with_type_condition_nullable.type;
+
+import com.apollographql.apollo.api.ScalarType;
+import java.lang.Class;
+import java.lang.Override;
+import java.lang.String;
+import javax.annotation.Generated;
+
+@Generated("Apollo GraphQL")
+public enum CustomType implements ScalarType {
+  ID {
+    @Override
+    public String typeName() {
+      return "ID";
+    }
+
+    @Override
+    public Class javaType() {
+      return String.class;
+    }
+  }
+}

--- a/apollo-compiler/src/test/kotlin/com/apollographql/apollo/compiler/CodegenTest.kt
+++ b/apollo-compiler/src/test/kotlin/com/apollographql/apollo/compiler/CodegenTest.kt
@@ -72,8 +72,9 @@ class CodeGenTest(val pkgName: String, val args: GraphQLCompiler.Arguments) {
             val nullableValueType = when {
               it.name == "hero_details_guava" -> NullableValueType.GUAVA_OPTIONAL
               it.name == "hero_details_java_optional" -> NullableValueType.JAVA_OPTIONAL
-              (it.name != "hero_details_nullable" || it.name == "no_accessors") -> NullableValueType.APOLLO_OPTIONAL
-              else -> NullableValueType.ANNOTATED
+              it.name == "fragments_with_type_condition_nullable" -> NullableValueType.ANNOTATED
+              it.name == "hero_details_nullable" -> NullableValueType.ANNOTATED
+              else -> NullableValueType.APOLLO_OPTIONAL
             }
             val useSemanticNaming = when {
               it.name == "hero_details_semantic_naming" -> true

